.travis.yml

```diff
@@ -17,21 +17,31 @@ language: python
 git:
   depth: 20
   quiet: true
+
+# Specify that the lint build stage should run first; if it fails, the test
+# build stage won't run.
+stages:
+  - lint
+  - test
+
+# This creates the lint stage, which only runs the linter. The relevant travis
+# feature is "Build stages" (https://docs.travis-ci.com/user/build-stages/).
+# Unfortunately it doesn't support matrix expansion, so we define the test
+# stage the "old-fashioned" way down below this section.
+jobs:
+  include:
+    - stage: lint
+      script: ./testing/run_lints.sh
+
+# The below implicitly run under a build stage called "Test". The full matrix
+# of python versions x SHARD will be expanded and run in parallel.
 python:
   - "2.7"
   - "3.6"
 env:
   global:
-    # We only test against the nightly version of TensorFlow.
-    - TF_VERSION="tf-nightly"
     - NUM_SHARDS=2
   matrix:
-    # We shard our tests to avoid timeouts. Additionally, only shard 0 runs the
-    # lints.
     - SHARD=0
     - SHARD=1
-install:
```

Comment:
I appreciate the reasoning behind not using this, but I still think a separate install script (placed in the script's section) has benefits. One specific benefit I have in mind is that in its log display, it'll collapse the outputs of individual scripts so we can be verbose as we want as in the install script. In particular, I like printing out the packages that pip installed so we can have an easier time debugging things if anything went wrong. Naturally, it also collapses the `set -x` spam.

Reply:
Hm, i don't think the collapsing thing is true for multiple scripts (i think it is for, e.g., separate install, before_script, etc sections) but i'll play around with it (tomorrow) and see if i can finagle some better collapsing. FWIW I believe both the apt and pip install commands will still print errors if anything goes wrong. If it's just to see what *was* installed, we could also add a 'pip freeze' or something. But I'm fine with the verbosity if I can figure out the collapsability.

Reply:
also someone on the internet points to [this](https://github.com/travis-ci/travis-rubies/blob/9f7962a881c55d32da7c76baefc58b89e3941d91/build.sh#L38-L44) as an example of how to add custom folds by hand. could give that a try

Reply:
ok tried it with the start_fold/end_fold commands. Here's an example output: https://travis-ci.org/csuter/probability/jobs/483687631

WDYT?

Reply:
That looks great, thanks!